### ISSUE TITLE
Netlify On-demand Builders - would they help with the Netlify Adapter?

### ISSUE THREAD

**Issue by benaltair** [sveltejs/kit#1286]
# Suggestion
Recently, Netlify released [a blog post](https://www.netlify.com/blog/2021/04/14/faster-builds-for-large-sites-on-netlify-with-on-demand-builders-now-in-early-access/) detailing On-demand Builders, a new feature that helps with SSR JS frameworks. It's in early access, but from my preliminary reading it seems it might help with Kit's [Netlify adapter](https://github.com/sveltejs/kit/tree/master/packages/adapter-netlify).

# Ways Forward
Perhaps On-demand Builders isn't compatible with Kit's current methodology, so this issue could be closed then. My hope was that in case it is compatible and helpful, soon it might be a good time to begin implementing it, while the project is young. Or maybe that would cause additional pain to have two pre-release projects interacting. Since it's only the adapter that should need to be modified, perhaps it's lower risk though.

I'm curious on any thoughts that come up. Feel free to close if it's not compatible with the project's stack or aims.

**Comment by sw-yx** [comment]
it is very early days for ODB's and imo Netlify needs to put out more documentation before we can assess if its ready for adoption. i think the ideas are inherently compatible, but probably need to introduce something like `prerender: onDemand` instead of a simple `true` or `false` boolean.

perhaps someone on @jlengstorf's team can help with implementing on demand builders?

**Comment by ascorbic** [comment]
Hey! I'm happy to help out with this. I see a few possibilities with the API, in additon to the one suggested by @sw-yx. 

First, there are the restrictions that need to be considered. ODBs are cached based on the URL, but without the query string. They only work for GET requests. For this reason they aren't suitable if you expect to generate different content based on query strings or headers. The cache lifetime is also not something you should rely on: they can be evicted for any reason, and are always evicted when a new build is done. For this reason they shouldn't be used for anything where the underlying data source may change. What they are great for though is long tail content, or other content that can change arbitrarily based on the URL.

One approach would be to apply it to all pages that are *not* pre-rendered, and leave regular functions for endpoints only. This wouldn't work if you expect pages to be updated between deploys, or if you expect to deliver different content based on anything except the URL. 
Alternatively there's @sw-yx 's approach of having a flag. My main concern about that is the detail: `prerender = onDemand` sounds like a contradiction! Maybe another flag `persist = true`, and don't set prerender. This also fits with the "[distributed persistent rendering](https://github.com/jamstack/jamstack.org/discussions/549)" name for the generic approach.

The main implementaiton issue is that ODBs are implemented by wrapping the render function, so it iether needs a separate entry point than other functions, or the wrapper needs some way of identifying an incoming request as on-demand so that it can pass it over to the ODB handler. We'd either need a way of seeing this in the request, or in the response generated by the app.

**Comment by sw-yx** [comment]
i just dislike filling up the sveltekit api with a bunch of boolean flags

**Comment by jlengstorf** [comment]
what about something like `prerender: deferred` or `prerender: serverless`? it sounds a little less contradictory _and_ avoids introducing new flags, while still lining up with the DPR reasoning (the same process as build-time prerendering is used, but the rendering is deferred until the first request)

**Comment by sw-yx** [comment]
yes, this would also work :) 

bottomline is i think sveltekit users are keen on trying out the ODB concept, they just need the kickstart of how to set this up/build this into the sveltekit netlify adapter

**Comment by benaltair** [comment]
@Xenonym I'm not entirely sure if this applies, but from what I understand in your comment, have you looked into dynamic redirects in Netlify? I wonder if this would apply here. Something with [placeholders](https://docs.netlify.com/routing/redirects/redirect-options/#placeholders) or `:splat` for instance could possibly pass the path data to a function as a query parameter or similar. Not familiar enough to know if that's a robust approach.

**Comment by sw-yx** [comment]
checking in on this again - @ascorbic i have a use case for this if you are keen to implement it in the sveltekit adapter https://github.com/sw-yx/swyxkit/issues/10

**Comment by serhalp** [comment]
@teemingc FYI we can close this one. On-Demand Builders were superseded years ago by Netlify's [second iteration on serverless functions](https://developers.netlify.com/guides/migrating-to-the-modern-netlify-functions/) and [headers-based caching](https://docs.netlify.com/build/caching/caching-overview/).

**Comment by teemingc** [comment]
Thanks! The header seems easy to add to the returned response in the user's `hooks.server.ts` file.
